src/amm.rs

```diff
@@ -28,31 +29,42 @@ macro_rules! to_u256 {
     );
 }
 
-macro_rules! to_u128 {
+macro_rules! to_balance {
     ($x:expr) => {
-        u128::try_from($x).ok()
+        Balance::try_from($x).map_err(|_| Overflow)
     };
 }
 
+#[derive(PartialEq)]
+#[derive(Debug)]
+pub enum MathError {
+    ZeroInReserve,
+    Overflow,
+    InsufficientOutReserve,
+}
+
 /// Calculate spot price given reserve of selling asset and reserve of buying asset.
 /// Formula : OUT_RESERVE * AMOUNT / IN_RESERVE
 ///
 /// - `in_reserve` - reserve amount of selling asset
 /// - `out_reserve` - reserve amount of buying asset
 /// - `amount` - amount
 ///
-/// Returns None in case of error
-pub fn calculate_spot_price(in_reserve: Balance, out_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(in_reserve != 0);
+/// Returns MathError in case of error
+pub fn calculate_spot_price(in_reserve: Balance, out_reserve: Balance, amount: Balance) -> Result<Balance, MathError> {
+    ensure!(in_reserve != 0, ZeroInReserve);
+
+    if amount == 0 || out_reserve == 0 {
+        return to_balance!(0);
+    }
 
     let (amount_hp, out_reserve_hp, in_reserve_hp) = to_u256!(amount, out_reserve, in_reserve);
 
     let spot_price_hp = out_reserve_hp
-        .checked_mul(amount_hp)
-        .expect("Cannot overflow")
-        .checked_div(in_reserve_hp)?;
+        .checked_mul(amount_hp).ok_or(Overflow).unwrap()
+        .checked_div(in_reserve_hp).ok_or(Overflow).unwrap();
```

Comment:
Here and everywhere where checked math is used and you do unwraps, you can remove the unwraps and change to:

`.checked_mul(amount_hp).ok_or(Overflow)?`

imo, it is much more concise.